src/core/server/metrics/integration_tests/server_collector.test.ts

```diff
@@ -0,0 +1,135 @@
+
+describe('ServerMetricsCollector', () => {
+  let server: HttpService;
+  let collector: ServerMetricsCollector;
+  let hapiServer: HapiServer;
+  let router: IRouter;
+
+  const delay = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
+  const sendGet = (path: string) => supertest(hapiServer.listener).get(path);
+
+  beforeEach(async () => {
+    server = createHttpServer();
+    const contextSetup = contextServiceMock.createSetupContract();
+    const httpSetup = await server.setup({ context: contextSetup });
+    hapiServer = httpSetup.server;
+    router = httpSetup.createRouter('/');
+    collector = new ServerMetricsCollector(hapiServer);
+  });
+
+  afterEach(async () => {
+    await server.stop();
+  });
+
+  it('collect requests infos', async () => {
+    router.get({ path: '/', validate: false }, async (ctx, req, res) => {
+      return res.ok({ body: '' });
+    });
+    await server.start();
+
+    let metrics = await collector.collect();
+
+    expect(metrics.requests).toEqual({
+      total: 0,
+      disconnects: 0,
+      statusCodes: {},
+    });
+
+    await sendGet('/');
+    await sendGet('/');
+    await sendGet('/not-found');
+
+    metrics = await collector.collect();
+
+    expect(metrics.requests).toEqual({
+      total: 3,
+      disconnects: 0,
+      statusCodes: {
+        '200': 2,
+        '404': 1,
+      },
+    });
+  });
+
+  it('collect response times', async () => {
+    router.get({ path: '/no-delay', validate: false }, async (ctx, req, res) => {
+      return res.ok({ body: '' });
+    });
+    router.get({ path: '/500-ms', validate: false }, async (ctx, req, res) => {
+      await delay(500);
+      return res.ok({ body: '' });
+    });
+    router.get({ path: '/250-ms', validate: false }, async (ctx, req, res) => {
+      await delay(250);
+      return res.ok({ body: '' });
+    });
+    await server.start();
+
+    await Promise.all([sendGet('/no-delay'), sendGet('/250-ms')]);
+    let metrics = await collector.collect();
+
+    expect(metrics.response_times.avg_in_millis).toBeGreaterThanOrEqual(125);
+    expect(metrics.response_times.max_in_millis).toBeGreaterThanOrEqual(250);
+
+    await Promise.all([sendGet('/500-ms'), sendGet('/500-ms')]);
+    metrics = await collector.collect();
+
+    expect(metrics.response_times.avg_in_millis).toBeGreaterThanOrEqual(250);
+    expect(metrics.response_times.max_in_millis).toBeGreaterThanOrEqual(500);
+  });
+
+  it.skip('collect connection count', async () => {
+    const waitSubject = new Subject();
+
+    router.get({ path: '/', validate: false }, async (ctx, req, res) => {
+      await waitSubject.pipe(take(1)).toPromise();
+      return res.ok({ body: '' });
+    });
+    await server.start();
+
+    let metrics = await collector.collect();
+    expect(metrics.concurrent_connections).toEqual(0);
+
+    sendGet('/');
+    await delay(20);
+    metrics = await collector.collect();
+    expect(metrics.concurrent_connections).toEqual(1);
```

Comment:
This is trying to test this part of the code:

```ts
    const connections = await new Promise<number>(resolve => {
      this.server.listener.getConnections((_, count) => {
        resolve(count);
      });
    });
```

I though `this.server.listener.getConnections` was returning the number of opened/pending connections, so I tried to test it by keeping waiting handlers, but the test fails ( `metrics.concurrent_connections` is always equals to 0). 

The snippet is directly copied from 

https://github.com/elastic/kibana/blob/8e9a8a84dccfa7965ce8a22362885e6cdef8b51f/src/legacy/server/status/index.js#L41-L49

So it's very unlikely is doesn't work (or at least doesn't do what it's supposed to do), however I don't know how to to properly integration test it ( I could unit test it by mocking `server.listener.getConnections`, but mocking the whole hapi server just for that felt like overkill)

If someone got an idea...

Reply:
`sendGet` doesn't send a request, you need to call `end` with a callback. https://visionmedia.github.io/superagent/
```js
    supertest(hapiServer.listener)
      .get('/')
      .end(() => null);
```

Reply:
It does when you `await` for it. using both `end` and awaiting it after throws an error stating that.

Reply:
Oh but I'm not awaiting for it here.... Thanks.